# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/ipc/RemoteAccessibleBase.cpp
# Commit: 54e0b1365f89
# Full Hash: 54e0b1365f898f89e7a72cd1297fcf48e43f37f5
# Author: Morgan Rae Reschenberg <mreschenberg@mozilla.com>
# Date: 2022-10-21 08:39:18
# Regressor Bug: 1787284
# File Overlap Count: 1
# Description:
#   Bug 1787284: [Part 3] Use cached name to compute MEMBER_OF rel for HTML radio buttons r=eeejay
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D159119
# ==============================================================================

diff -r c6db418af190 -r 54e0b1365f89 accessible/ipc/RemoteAccessibleBase.cpp
--- a/accessible/ipc/RemoteAccessibleBase.cpp	Thu Oct 20 21:40:06 2022 +0000
+++ b/accessible/ipc/RemoteAccessibleBase.cpp	Thu Oct 20 21:40:06 2022 +0000
@@ -726,6 +726,28 @@
 
   if (aType == RelationType::MEMBER_OF) {
     Relation rel = Relation();
+    // HTML radio buttons with cached names should be grouped.
+    if (IsHTMLRadioButton()) {
+      auto maybeName =
+          mCachedFields->GetAttribute<nsString>(nsGkAtoms::radioLabel);
+      if (!maybeName) {
+        return rel;
+      }
+
+      RemoteAccessible* ancestor = RemoteParent();
+      while (ancestor && ancestor->Role() != roles::FORM && ancestor != mDoc) {
+        ancestor = ancestor->RemoteParent();
+      }
+      Pivot p = Pivot(ancestor);
+      PivotRadioNameRule rule(*maybeName);
+      Accessible* match = p.Next(ancestor, rule);
+      while (match) {
+        rel.AppendTarget(match->AsRemote());
+        match = p.Next(match, rule);
+      }
+      return rel;
+    }
+
     if (IsARIARole(nsGkAtoms::radio)) {
       // ARIA radio buttons should be grouped by their radio group
       // parent, if one exists.
